File: firmware/stampfly/tests/test_flight_link_protocol.cpp
#include "flight_link_protocol.hpp"

#include <assert.h>
#include <stdint.h>

using namespace flight_link;

static SetCommand nonzero_set() {
    SetCommand command;
    command.roll = 0.25f;
    command.pitch = -0.5f;
    command.yaw = 0.125f;
    command.throttle = 0.4f;
    return command;
}

int main() {
    const uint64_t session_id = 0x0102030405060708ULL;
    const Packet original = make_set(session_id, 2, 250, nonzero_set());
    std::vector<uint8_t> wire;
    assert(encode(original, &wire) == Error::NONE);
    assert(wire.size() == kHeaderBytes + 10 + kCrcBytes);
    DecodeResult decoded = decode(wire.data(), wire.size());
    assert(decoded.ok());
    SetCommand parsed;
    assert(decode_set(decoded.packet, &parsed) == Error::NONE);
    assert(parsed.roll == 0.25f);
    assert(parsed.pitch == -0.5f);
    assert(parsed.control_mode == 0);

    wire[wire.size() - 1] ^= 0x01;
    assert(decode(wire.data(), wire.size()).error == Error::CHECKSUM);
    assert(decode(nullptr, 0).error == Error::INVALID_LENGTH);

    SessionState state;
    assert(state.accept(make_claim(session_id, 1, 250), 1000) == Error::NONE);
    assert(state.claimed());
    assert(state.disarmed());
    assert(state.accept(original, 1010) == Error::NONE);
    assert(!state.disarmed());
    assert(state.connected(1260));
    assert(state.watchdog_expired(1261));

    const Packet duplicate = original;
    assert(state.accept(duplicate, 1011) == Error::DUPLICATE_SEQUENCE);
    assert(state.last_control_ms() == 1010);
    assert(state.accept(make_set(session_id, 1, 250, SetCommand{}), 1012) == Error::STALE_SEQUENCE);
    assert(state.accept(make_set(session_id + 1, 3, 250, SetCommand{}), 1013) == Error::WRONG_SESSION);
    assert(state.accept(make_claim(session_id, 1, 250), 1014) == Error::STALE_SEQUENCE);
    assert(!state.disarmed());
    assert(state.accept(make_claim(session_id, 3, 250), 1015) == Error::NONE);
    assert(state.disarmed());

    const Packet application_ack = make_ack(session_id, 4, 250, MessageKind::SET, 2, AckClass::APPLICATION, true);
    const Packet delivery = make_ack(session_id, 5, 250, MessageKind::SET, 2, AckClass::RF_DELIVERY, true);
    assert(state.accept_application_ack(application_ack, 1014) == Error::NONE);
    assert(state.accept_application_ack(delivery, 1015) == Error::ACK_NOT_APPLICABLE);
    assert(state.last_control_ms() == 1015);
    assert(state.accept(application_ack, 1016) == Error::ACK_NOT_APPLICABLE);

    SessionState exhaustion;
    assert(exhaustion.accept(make_claim(session_id, 1, 250), 0) == Error::NONE);
    assert(exhaustion.accept(make_set(session_id, kMaxSequence, 250, SetCommand{}), 1) == Error::NONE);
    assert(exhaustion.sequence_exhausted());
    assert(exhaustion.accept(make_set(session_id, 3, 250, SetCommand{}), 2) == Error::SEQUENCE_EXHAUSTED);
    assert(exhaustion.accept(make_claim(session_id + 1, 1, 250), 3) == Error::NONE);
    assert(exhaustion.disarmed());

    assert(encode(make_set(session_id, 4, 0, SetCommand{}), &wire) == Error::INVALID_TTL);
    SetCommand invalid = SetCommand{};
    invalid.roll = 2.0f;
    assert(encode(make_set(session_id, 5, 250, invalid), &wire) == Error::INVALID_PAYLOAD);
    return 0;
}

File: host/tests/test_flight_link_protocol.py
from __future__ import annotations

import unittest

from host.flight_link_protocol import (
    AckClass,
    Error,
    FlightLinkError,
    MessageKind,
    SessionState,
    SetCommand,
    decode,
    decode_ack,
    decode_set,
    encode,
    make_ack,
    make_claim,
    make_set,
)


class FlightLinkProtocolTests(unittest.TestCase):
    def test_round_trip_and_ack_classes(self):
        command = SetCommand(0.25, -0.5, 0.125, 0.4)
        packet = make_set(0x0102030405060708, 2, 250, command)
        decoded = decode(encode(packet))
        self.assertEqual(decode_set(decoded), command)

        application = decode_ack(
            decode(encode(make_ack(1, 3, 250, MessageKind.SET, 2, AckClass.APPLICATION, True)))
        )
        delivery = decode_ack(
            decode(encode(make_ack(1, 4, 250, MessageKind.SET, 2, AckClass.RF_DELIVERY, True)))
        )
        self.assertEqual(application.ack_class, AckClass.APPLICATION)
        self.assertEqual(delivery.ack_class, AckClass.RF_DELIVERY)

    def test_malformed_packet_is_rejected_before_state(self):
        packet = bytearray(encode(make_claim(1, 1, 250)))
        packet[-1] ^= 1
        with self.assertRaises(FlightLinkError) as context:
            decode(bytes(packet))
        self.assertEqual(context.exception.error, Error.CHECKSUM)

        state = SessionState()
        self.assertEqual(state.accept(make_claim(1, 1, 250), 1000), Error.NONE)
        self.assertEqual(state.last_control_ms, 1000)
        malformed = make_set(1, 2, 250, SetCommand())
        malformed = type(malformed)(
            malformed.version,
            malformed.kind,
            malformed.session_id,
            malformed.sequence,
            malformed.ttl_ms,
            malformed.capabilities,
            b"bad",
        )
        self.assertEqual(state.accept(malformed, 1001), Error.INVALID_PAYLOAD)
        self.assertEqual(state.last_sequence, 1)
        self.assertEqual(state.last_control_ms, 1000)

    def test_duplicate_reordered_wrong_session_expiry_and_reconnect(self):
        state = SessionState()
        self.assertEqual(state.accept(make_claim(9, 1, 250), 0xFFFFFF00), Error.NONE)
        nonzero = make_set(9, 2, 250, SetCommand(throttle=0.4))
        self.assertEqual(state.accept(nonzero, 0xFFFFFF10), Error.NONE)
        self.assertFalse(state.disarmed)
        self.assertEqual(state.accept(nonzero, 0xFFFFFF11), Error.DUPLICATE_SEQUENCE)
        self.assertEqual(state.accept(make_set(9, 1, 250, SetCommand()), 0xFFFFFF12), Error.STALE_SEQUENCE)
        self.assertEqual(state.accept(make_set(10, 3, 250, SetCommand()), 0xFFFFFF13), Error.WRONG_SESSION)
        self.assertTrue(state.connected(0xFFFFFFF0))
        self.assertTrue(state.watchdog_expired(0x00000010))

        self.assertEqual(state.accept(make_claim(9, 1, 250), 0xFFFFFFF1), Error.STALE_SEQUENCE)
        self.assertFalse(state.disarmed)
        self.assertEqual(state.accept(make_claim(9, 3, 250), 0xFFFFFFF2), Error.NONE)
        self.assertTrue(state.disarmed)

        self.assertEqual(state.accept(make_claim(10, 1, 250), 20), Error.NONE)
        self.assertTrue(state.disarmed)
        self.assertEqual(state.setpoint, SetCommand())

    def test_set_range_and_ttl_fail_closed(self):
        with self.assertRaises(FlightLinkError):
            encode(make_set(1, 1, 0, SetCommand()))
        with self.assertRaises(FlightLinkError):
            make_set(1, 1, 250, SetCommand(roll=2.0))


if __name__ == "__main__":
    unittest.main()
